optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/uni/UniConstraintStream.java

```diff
@@ -186,20 +186,87 @@
+     * @param <GroupKey_> the type of a fact in the destination {@link UniConstraintStream}'s tuple
+     * @return never null
+     */
+    <GroupKey_> UniConstraintStream<GroupKey_> groupBy(Function<A, GroupKey_> groupKeyMapping);
+
+    /**
+     * Expands the {@link UniConstraintStream} to a {@link BiConstraintStream}, consisting of tuples which have:
+     *
+     * <ul>
+     *     <li>As the first fact, the value resulting from applying the group key mapping function on the fact from the
+     *     original tuple.
+     *     Neither such facts of the new stream will {@link Objects#equals(Object, Object)} any other.</li>
+     *     <li>As the second fact, the value of applying a given {@link UniConstraintCollector} on all the facts of the
+     *     original tuple with a matching group key.</li>
+     * </ul>
+     *
+     * @param groupKeyMapping never null, function to convert a fact in original tuple to a different fact
+     * @param <GroupKey_> the type of the first fact in the destination {@link BiConstraintStream}'s tuple
+     * @param <ResultContainer_> the mutable accumulation type (often hidden as an implementation detail)
+     * @param <Result_> the type of the second fact in the destination {@link BiConstraintStream}'s tuple
+     * @return never null
+     */
     <GroupKey_, ResultContainer_, Result_> BiConstraintStream<GroupKey_, Result_> groupBy(
             Function<A, GroupKey_> groupKeyMapping,
             UniConstraintCollector<A, ResultContainer_, Result_> collector);
 
+    /**
+     * Expand the {@link UniConstraintStream} to a {@link BiConstraintStream}, consisting of tuples which have:
+     *
+     * <ul>
+     *     <li>As the first fact, the value resulting from applying the "A" group key mapping function to the fact from
+     *     the original tuple.</li>
+     *     <li>As the second fact, the value resulting from applying the "B" group key mapping function to the fact from
+     *     the original tuple, which already passed through the "A" group key mapping.</li>
+     * </ul>
+     *
+     * All bi tuples of the new stream will be unique.
+     *
+     * @param groupKeyAMapping never null, function to convert first fact in the original tuple to a different fact
+     * @param groupKeyBMapping never null, function to convert second fact in the original tuple to a different fact
+     * @param <GroupKeyA_> the type of the first fact in the destination {@link BiConstraintStream}'s tuple
+     * @param <GroupKeyB_> the type of the second fact in the destination {@link BiConstraintStream}'s tuple
+     * @return never null
+     */
     <GroupKeyA_, GroupKeyB_> BiConstraintStream<GroupKeyA_, GroupKeyB_> groupBy(
             Function<A, GroupKeyA_> groupKeyAMapping,
             Function<A, GroupKeyB_> groupKeyBMapping);
 
+    /**
+     * Merges the semantics of {@link #groupBy(Function, Function)} and {@link #groupBy(UniConstraintCollector)}.
```

Comment:
I'd avoid the use of the word "merge" here (it confuses to think that the streams are merged).
It could be "Combines the semantics ...", but maybe we can leave it out entirely:
"A groupBy with two group key and 1 collector that leads to a TriConstraintStream."